library_generation/utilities.sh

```diff
@@ -0,0 +1,178 @@
+#!/usr/bin/env bash
+
+set -xeo pipefail
+
+# define utility functions
+
+extract_folder_name() {
+  destination_path=$1
+  folder_name=${destination_path##*/}
+  echo "$folder_name"
+}
+
+remove_empty_files() {
+  category=$1
+  find "$destination_path/$category-$folder_name/src/main/java" -type f -size 0 | while read -r f; do rm -f "${f}"; done
+  if [ -d "$destination_path/$category-$folder_name/src/main/java/samples" ]; then
+      mv "$destination_path/$category-$folder_name/src/main/java/samples" "$destination_path/$category-$folder_name"
+  fi
+}
+
+# Move generated files to folders in destination_path.
+mv_src_files() {
+  category=$1 # one of gapic, proto, samples
+  type=$2 # one of main, test
+  if [ "$category" == "samples" ]; then
+    src_suffix="samples/snippets/generated/src/main/java/com"
+    folder_suffix="samples/snippets/generated"
+  elif [ "$category" == "proto" ]; then
+    src_suffix="$category/src/$type/java"
+    folder_suffix="$category-$folder_name/src/$type"
+  else
+    src_suffix="src/$type"
+    folder_suffix="$category-$folder_name/src"
+  fi
+  mkdir -p "$destination_path/$folder_suffix"
+  cp -r "$destination_path/java_gapic_srcjar/$src_suffix" "$destination_path/$folder_suffix"
+  if [ "$category" != "samples" ]; then
+    rm -r -f "$destination_path/$folder_suffix/java/META-INF"
+  fi
+}
+
+# unzip jar file
+unzip_src_files() {
+  category=$1
+  jar_file=java_$category.jar
+  mkdir -p "$destination_path/$category-$folder_name/src/main/java"
+  unzip -q -o "$destination_path/$jar_file" -d "$destination_path/$category-$folder_name/src/main/java"
+  rm -r -f "$destination_path/$category-$folder_name/src/main/java/META-INF"
+}
+
+find_additional_protos_in_yaml() {
+  pattern=$1
+  find_result=$(grep --include=\*.yaml -rw "$proto_path" -e "$pattern")
+  if [ -n "$find_result" ]; then
+    echo "$find_result"
+  fi
+}
+
+# Apart from proto files in proto_path, additional protos are needed in order
+# to generate gapic client libraries.
+# Search additional protos in .yaml files.
+search_additional_protos() {
```

Comment:
This whole search for additional protos seems to assume that the protos are always coming from `googleapis`. Can we instead explicitly add a `generate_library.sh` parameter for locations of additional protos?

Reply:
In bazel, all protos are listed as dependencies. 

We can infer that list by downloading `google` folder from `googleapis`. Can I assume all proto files are coming from `google` folder?

In design, we want the script can generate libraries with only the given set of protos, but other protos (e.g., `googleapis/google/api`) are needed to generate a library (showcase is an example), so I think we can't get rid of `googleapis` entirely.

Reply:
Follow up on meetings and chat, I added instructions on users should pull `google/` directory from `googleapis` before running the script. These additional protos can be found within `google/` directory.

Reply:
[Related](https://github.com/googleapis/sdk-platform-java/pull/1935/files#r1301828838)

Reply:
I think it's OK to have some assumptions at this moment if we make them clear. However it's not very clear to me that where we should put these common protos, and where we should run the script. Can we run the script from anywhere and point `poto_path` to any folder? Left some comments on the README in respective sections.

Reply:
We can run `generate_library.sh` anywhere as long as the script knows how to find all protos (not only protos listed in `proto_path`, but also protos referenced by those protos).

Since we don't know whether we have the permission of an arbitrary directory, it's safe to set `proto_path` as a relative path in the current working directory.

Reply:
> as long as the script knows how to find all protos

What does this mean? How do we let the script know how to find protos? Does the folder structure has to be in certain ways?

Reply:
> What does this mean? How do we let the script know how to find protos? Does the folder structure has to be in certain ways?

By using `proto_path`. 

The folder structure should be the same as the import statements in protos.

For example, we want to generate from `folder1/folder2/protoA`, so `proto_path` should be set to `folder1/folder2`. protoA imports protoB as `folder3/folder4/protoB`, then there should be `folder3/folder4/` (containing `protoB`) in the current working directory (`$cwd`, the directory in which `generate_library.sh` resides).

When generating protoA, protoc will first try to generate protoB (because protoA imports protoB) and it will search for protoB in `folder3/folder4` (notice the relative path). If protoB doesn't in `folder3/folder4`, protoc will exit with error.

Reply:
Thanks, that's much clearer! Can you please add these explanation to README somewhere?

Reply:
I added an explanation in prerequisite ([commit](https://github.com/googleapis/sdk-platform-java/pull/1916/commits/1648254a8ab5e10312aa93fc0c0970ccc10a9d6c)).